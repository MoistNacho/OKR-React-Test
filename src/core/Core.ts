import { observable } from "mobx";

import Dialog from "./Dialog";
import PopUp from "./PopUp";
import Router from "./Router";

export default class Core {
  public dialog: Dialog;

  public router: Router;

  public popUp: PopUp;

  @observable
  public locale = "ko";

  private stores: {
    [key: string]: any;
  } = {};

  constructor() {
    this.dialog = new Dialog();
    this.router = new Router();
    this.popUp = new PopUp();

    // expose core during tests
    if ((window as any).Cypress) {
      (window as any).core = this;
    }
  }

  public addStore = (key: string, store: any) => {
    this.stores[key] = store;
  };

  public getStore = (key: string) => this.stores[key];

  public clearStores = () => {
    this.stores = {};
  };

  public reset = () => this.clearStores();
}
