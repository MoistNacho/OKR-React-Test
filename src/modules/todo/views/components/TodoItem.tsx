import { ButtonV2 } from "@meshkorea/vroong-design-system-web";
import React, { useCallback, useRef, useState } from "react";
import styled from "styled-components";

import { TodoItem } from "modules/todo/TodoStore";

interface TodoItemProps {
  item: TodoItem;
  id: number;
  onUpdate: (id: number, value: string) => void;
  onRemove: (id: number) => void;
}

const TodoItem = ({ item, id, onUpdate, onRemove }: TodoItemProps) => {
  const [isEdit, setIsEdit] = useState(false);
  const [inputError, setInputError] = useState(false);
  const inputRef = useRef<HTMLInputElement>(null);

  const handleEdit = useCallback(() => {
    setIsEdit(!isEdit);
    setInputError(false);
  }, [isEdit]);

  const handleUpdate = useCallback(() => {
    const { value } = inputRef.current!;

    if (!value) {
      setInputError(true);
    } else {
      onUpdate(item.id, value);
      setIsEdit(false);
      setInputError(false);
    }
  }, [item.id, onUpdate]);

  const handleRemove = useCallback(() => {
    onRemove(item.id);
  }, [item.id, onRemove]);

  return (
    <ListItem>
      {inputError && <Required>아이템 이름은 필수값입니다</Required>}
      {isEdit ? (
        <InputWrap>
          <input
            ref={inputRef}
            type="text"
            placeholder="아이템 이름을 입력해주세요"
            defaultValue={item.name}
          />
          <ButtonsWrap>
            <ButtonV2 onClick={handleUpdate}>완료</ButtonV2>
            <ButtonV2 style={{ color: "#fa5c5c" }} onClick={handleEdit}>
              취소
            </ButtonV2>
          </ButtonsWrap>
        </InputWrap>
      ) : (
        <InputWrap>
          <span>{`${id}. ${item.name}`}</span>
          <ButtonsWrap>
            <ButtonV2 onClick={handleEdit}>수정</ButtonV2>
            <ButtonV2 onClick={handleRemove} style={{ color: "#fa5c5c" }}>
              삭제
            </ButtonV2>
          </ButtonsWrap>
        </InputWrap>
      )}
    </ListItem>
  );
};

export default TodoItem;

const ListItem = styled.li`
  color: ${(props) => props.theme.palette.text.SecondaryText};
`;

const InputWrap = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const ButtonsWrap = styled.div`
  display: flex;
`;

const Required = styled.p`
  font-size: 10px;
  color: #ff4949;
  margin: 0 0 4px;

  ::before {
    display: inline;
    content: "⚠ ";
  }
`;
