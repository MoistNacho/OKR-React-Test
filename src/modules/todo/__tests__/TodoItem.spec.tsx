import { BaseTheme } from "@meshkorea/vroong-design-system-web";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import React from "react";
import renderer from "react-test-renderer";
import { ThemeProvider } from "styled-components";

import { CoreProvider } from "core";

import TodoItem from "../views/components/TodoItem";

describe("<TodoItem />", () => {
  it("renders with snapshot", () => {
    const component = renderer.create(
      <CoreProvider>
        <ThemeProvider theme={BaseTheme}>
          <TodoItem
            item={{ id: 0, name: "renderTest" }}
            id={1}
            onUpdate={jest.fn()}
            onRemove={jest.fn()}
          />
        </ThemeProvider>
      </CoreProvider>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe("Update and Remove Test", () => {
    const onUpdate = jest.fn();
    const onRemove = jest.fn();
    let input: HTMLInputElement;
    let editButton: HTMLButtonElement;
    let updateButton: HTMLButtonElement;
    let removeButton: HTMLButtonElement;

    beforeEach(() => {
      render(
        <CoreProvider>
          <ThemeProvider theme={BaseTheme}>
            <TodoItem
              item={{ id: 0, name: "Test Item" }}
              id={1}
              onUpdate={onUpdate}
              onRemove={onRemove}
            />
          </ThemeProvider>
        </CoreProvider>,
      );
      editButton = screen.getByRole("button", { name: "수정" });
      removeButton = screen.getByRole("button", { name: "삭제" });
    });

    // it("'수정'버튼을 클릭시 '완료' 버튼으로 교체된다", async () => {
    //   await userEvent.click(editButton);
    //   updateButton = screen.getByRole("button", { name: "완료" });

    //   expect(updateButton).toBeInTheDocument();
    // });

    describe("Update", () => {
      beforeEach(async () => {
        await userEvent.click(editButton);
        input = screen.getByPlaceholderText("아이템 이름을 입력해주세요");
        updateButton = screen.getByRole("button", { name: "완료" });
      });

      it("'완료' 버튼을 클릭시 아이템의 이름이 비어있을 경우 오류메시지가 출력된다", async () => {
        await userEvent.clear(input);
        await userEvent.click(updateButton);
        const errorMessage = screen.getByText("아이템 이름은 필수값입니다");

        expect(errorMessage).toBeInTheDocument();
      });

      it("'완료' 버튼을 클릭시 아이템의 값과 함께 onUpdate 함수가 호출된다", async () => {
        await userEvent.clear(input);
        await userEvent.type(input, "New World");
        await userEvent.click(updateButton);

        // onUpdate함수는 item객체의 id값과 name값을 인자로 받는다
        expect(onUpdate).toHaveBeenCalledWith(0, "New World");
      });
    });

    it("'삭제' 버튼을 클릭시 아이템의 id값과 함께 onRemove 함수가 호출된다", async () => {
      await userEvent.click(removeButton);

      expect(onRemove).toHaveBeenCalledWith(0);
    });
  });
});
