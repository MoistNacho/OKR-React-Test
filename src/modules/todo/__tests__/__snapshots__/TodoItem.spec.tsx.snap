// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoItem /> renders with snapshot 1`] = `
<li
  className="sc-bPPhlf kkCVUQ"
>
  <div
    className="sc-hrZiYQ eyhTWK"
  >
    <span>
      1. renderTest
    </span>
    <div
      className="sc-caiQmJ cMBdbM"
    >
      <button
        className="sc-bczRLJ iehTtZ"
        disabled={false}
        onClick={[Function]}
        onKeyPress={[Function]}
        type="button"
      >
        수정
      </button>
      <button
        className="sc-bczRLJ iehTtZ"
        disabled={false}
        onClick={[Function]}
        onKeyPress={[Function]}
        style={
          Object {
            "color": "#fa5c5c",
          }
        }
        type="button"
      >
        삭제
      </button>
    </div>
  </div>
</li>
`;
